site/src/components/SelectMenu/SelectMenu.tsx

```diff
@@ -0,0 +1,140 @@
+import CheckOutlined from "@mui/icons-material/CheckOutlined";
+import Button, { type ButtonProps } from "@mui/material/Button";
+import MenuItem, { type MenuItemProps } from "@mui/material/MenuItem";
+import MenuList, { type MenuListProps } from "@mui/material/MenuList";
+import {
+  type FC,
+  forwardRef,
+  Children,
+  isValidElement,
+  type HTMLProps,
+} from "react";
+import { DropdownArrow } from "components/DropdownArrow/DropdownArrow";
+import {
+  Popover,
+  PopoverContent,
+  PopoverTrigger,
+} from "components/Popover/Popover";
+import {
+  SearchField,
+  type SearchFieldProps,
+} from "components/SearchField/SearchField";
+
+const SIDE_PADDING = 16;
+
+export const SelectMenu = Popover;
+
+export const SelectMenuTrigger = PopoverTrigger;
+
+export const SelectMenuContent = PopoverContent;
+
+export const SelectMenuButton = forwardRef<HTMLButtonElement, ButtonProps>(
+  (props, ref) => {
+    return (
+      <Button
+        css={{
+          // Icon and text should be aligned to the left
+          justifyContent: "flex-start",
+          flexShrink: 0,
+          "& .MuiButton-startIcon": {
+            marginLeft: 0,
+            marginRight: SIDE_PADDING,
+          },
+          // Dropdown arrow should be at the end of the button
+          "& .MuiButton-endIcon": {
+            marginLeft: "auto",
+          },
+        }}
+        endIcon={<DropdownArrow />}
+        ref={ref}
+        {...props}
+      >
+        <span
+          // Make sure long text does not break the button layout
+          css={{
+            display: "block",
+            overflow: "hidden",
+            textOverflow: "ellipsis",
+          }}
+        >
+          {props.children}
+        </span>
+      </Button>
+    );
+  },
+);
+
+export const SelectMenuSearch: FC<SearchFieldProps> = (props) => {
+  return (
+    <SearchField
+      fullWidth
+      size="medium"
+      css={(theme) => ({
+        "& input": {
+          fontSize: 14,
+        },
+
+        "& fieldset": {
+          border: 0,
+          borderRadius: 0,
+          borderBottom: `1px solid ${theme.palette.divider} !important`,
+        },
+        "& .MuiInputBase-root": {
+          padding: `12px ${SIDE_PADDING}px`,
+        },
+        "& .MuiInputAdornment-positionStart": {
+          marginRight: SIDE_PADDING,
+        },
+      })}
+      {...props}
+    />
+  );
+};
+
+export const SelectMenuList: FC<MenuListProps> = (props) => {
+  const items = Children.toArray(props.children);
+  type ItemType = (typeof items)[number];
+  const selectedAsFirst = (a: ItemType, b: ItemType) => {
+    if (
+      !isValidElement<MenuItemProps>(a) ||
+      !isValidElement<MenuItemProps>(b)
+    ) {
+      throw new Error(
+        "SelectMenuList children must be SelectMenuItem components",
+      );
+    }
+    return a.props.selected ? -1 : 0;
```

Comment:
Not sure if I got it, can you explain that using an example?

Reply:
Sure – this is from[ the MDN page on `Array.sort`'s callback](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#comparefn):
> It should return a number where:
> - A negative value indicates that a should come before b.
> - A positive value indicates that a should come after b.
> - Zero or NaN indicates that a and b are considered equal.

The current function only deals with 0 and -1, though. So we have no way of moving `a` after `b` – we're only ever able to move `a` before `b`, or keep them in the same position. At best, that means that are sorting is inefficient. If `a` isn't selected, but `b` is, we have no way of moving `a` after `b`. We just have to hope and pray that at some point during the comparisons, the sorting algorithm will compare `a` and `b` the opposite way, and then we can move the selected element ahead

But Array.sort won't stop until the array is fully sorted. So if we're missing 1/3 of our operations, there's a risk that the function won't ever be able to fully sort the array. And because this sort is happening in a render, that means that the React UI itself will hang forever, because the sort will block everything

Reply:
Now that I write that out, though, I do think we should update the sort logic, just to be extra safe. This should do the trick:

```tsx
if (a.props.selected && b.props.selected) {
  return 0;
}

return a.props.selected ? -1 : 1;

Reply:
I think you meant `a.props.selected === b.props.selected`. As is, this would return `1` if both were `false`.

Reply:
I decided to take a diff approach than sorting 👍